parley/src/layout/editor.rs

```diff
@@ -174,6 +176,8 @@ where
 
     /// Delete the selection or the previous cluster (typical ‘backspace’ behavior).
     pub fn backdelete(&mut self) {
+        debug_assert!(!self.editor.is_composing());
```

Comment:
You are liable to end up in a really janky state.

If you really want us to implement a fallback, then making them a no-op would be the right one. It is, as documented, a logic error to call them, and they can't have any meaningful semantics. Doing an early return would be some quite ugly boilerplate:
```rust
if self.editor.is_composing() {
    if cfg!(debug_assertions) {
         panic!("Can't call most PlainEditor methods whilst composing");
    } else {
         return;
    }
}
```
But if you want us to the check unconditionally, we may as well just make it a full `assert`.

Cancelling compose mode *isn't something we can do* at this level of semantics; to cancel compose mode, you need to tell that platform that you're doing so, and we don't know what the platform is.

Reply:
> You are liable to end up in a really janky state.
> If you really want us to implement a fallback, then making them a no-op would be the right one.

My main point here being that if debug assertions are disabled, then the current code will lead to any such janky state (because there is no protection other than the debug assertion). Seems to me it should either be an unconditional panic or an early return noop (perhaps also logging a warning). I'd probably prefer the latter as I don't tend to like libraries panicking on me unexpectedly in produciton.

> Doing an early return would be some quite ugly boilerplate:

Perhaps a good case for a macro_rules macro?

Reply:
I'd be happy with it being a noop or assert on release, and an assert in debug mode. The janky state can get as bad as leading to crashes when the preedit byte offsets go out of text buffer bounds.